src/ray/util/concurrent_flat_map.h

```diff
@@ -0,0 +1,150 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/util/mutex_protected.h"
+
+namespace ray {
+
+// A flat map that's protected by a single mutex with read/write locks.
+//
+// Use when expecting one of the following:
+// - mininal thread contention
+// - trivially copyable values
+// - majority reads
+// If these don't describe your use case, prefer boost::concurrent_flat_map.
+template <typename KeyType, typename ValueType>
+class ConcurrentFlatMap {
+ public:
+  template <typename... Args>
+  explicit ConcurrentFlatMap(Args &&...args) : map_(std::forward(args)...) {}
+
+  ConcurrentFlatMap() = default;
+
+  void Reserve(size_t size) {
+    auto write_lock = map_.LockForWrite();
+    write_lock.Get().reserve(size);
+  }
+
+  template <typename KeyLike>
+  std::optional<ValueType> Get(const KeyLike &key) const {
+    std::optional<ValueType> result;
+    auto read_lock = map_.LockForRead();
+    const auto &map = read_lock.Get();
+    const auto it = map.find(key);
+    if (it != map.end()) {
+      result.emplace(it->second);
+    }
+    return result;
+  }
+
+  // Returns true if key existed and visitor was called.
+  // Visitor is called under a write lock so should not be heavy, otherwise prefer Get
+  // followed by InsertOrAssign.
+  template <typename KeyLike>
+  bool WriteVisit(const KeyLike &key, const std::function<void(ValueType &)> &visitor) {
+    auto write_lock = map_.LockForWrite();
+    auto &map = write_lock.Get();
+    const auto it = map.find(key);
+    if (it != map.end()) {
+      visitor(it->second);
+      return true;
+    }
+    return false;
+  }
+
+  // Returns true if key existed and visitor was called.
+  // Calls function under read lock so should not be heavy, otherwise prefer Get to copy
+  // out.
+  template <typename KeyLike>
+  bool ReadVisit(const KeyLike &key,
+                 const std::function<void(const ValueType &)> &visitor) const {
+    auto read_lock = map_.LockForRead();
+    const auto &map = read_lock.Get();
+    auto it = map.find(key);
+    if (it != map.end()) {
+      visitor(it->second);
+      return true;
+    }
+    return false;
+  }
+
+  // Applies visitor to all values in the map.
+  // Calls under read lock so should not be heavy, otherwise prefer GetMapClone to copy.
+  void ReadVisitAll(
```

Comment:
BTW, you could take templated visitor here.

Reply:
Example:
```
#include "absl/container/flat_hash_map.h"
#include <string>
#include <string_view>
#include "absl/types/span.h"
#include <iostream>
#include <vector>

absl::flat_hash_map<std::string, std::string> m;

template <typename Visitor>
void MyFunc(Visitor&& visitor) {
    std::forward<Visitor>(visitor)(m.begin()->first, m.begin()->second);
}

int main() {
    m.emplace("hello", "world");

    auto f = [](std::string_view key, std::string_view val) {
        std::cout << "key = " << key << std::endl;
        std::cout << "val = " << val << std::endl;
    };
    MyFunc(std::move(f));

    return 0;
}
```

Reply:
Usually we use `std::function` for type erasure because we want to have a unified type as data member, but not the case here.

In the example above, we could
- save copy for key and value pair
- save heap allocation in the `std::function` (especially you capture everything at the caller side...)